aip/client-libraries/4290.md

```diff
@@ -0,0 +1,202 @@
+---
+aip:
+  id: 4290
+  scope: client-libraries
+  state: approved
+  created: 2018-11-08
+  updated: 2019-05-29
+permalink: /client-libraries/4290
+redirect_from:
+  - /4290
+---
+
+# Docker interface
+
+A consequence of using individual generators for API client library generation
+is that each generator has its own set of dependencies and requirements in
+order to run.
+
+This is reasonable for a user who wishes to generate many libraries for a
+single environment, but presents challenges for a user wishing to generate a
+single API for many languages or environments. Users need a way to generate
+libraries easily and quickly, with minimal ramp-up per language.
+
+## Guidance
+
+Client library generators **should** ship [Docker][] images providing the
+generator and exposing a common interface, so that generating the same API in
+multiple languages usually only requires substituting in the appropriate Docker
+image. Docker images for Google-authored generators will follow a consistent
+scheme.
+
+### CLI usage
+
+The expected user command to invoke the code generator in a Docker image (from
+the proto import root, on a POSIX machine):
+
+```bash
+$ docker run --rm --user $UID \
+  --mount type=bind,source=a/b/c/v1/,destination=/in/a/b/c/v1/,readonly \
+  --mount type=bind,source=dest/,destination=/out/ \
+  gcr.io/gapic-images/{GENERATOR} \
+  [-- additional options...]
+```
+
+If a generator requires additional arguments (e.g. the Go GAPIC package
+argument), these should be added as arguments at the end (the last line in the
+example above), and a Docker entrypoint **must** route them to the appropriate
+`--{lang}_gapic_opt` arguments. However, generators **should not** require
+arguments, and this should be a last resort if they are unavoidable.
```

Comment:
Maybe add something like: `See [Plugin Options][] for more details.`

Reply:
I opted to remove this paragraph since it is repetitive.